From b91c5a1253279d0111c5ffb404695078d48f48b2 Mon Sep 17 00:00:00 2001
From: Corentin Oparowski <corentin.oparowski@vates.tech>
Date: Fri, 18 Jul 2025 18:02:35 +0800
Subject: [PATCH 1/8] CP-46343: common data structure padding

Signed-off-by: Chunjie Zhu <chunjie.zhu@cloud.com>

[ Corentin: rebase from XS-9 kernel (v6.6.98 based) to UEK8 kernel (v6.12 based):
Replaced every instance of xs_kabi_padding by  UEK_KABI_RESERVE(1-4)]
---
 block/elevator.h                    | 11 +++++++++++
 drivers/char/ipmi/ipmi_si.h         |  6 ++++++
 drivers/char/ipmi/ipmi_si_sm.h      |  6 ++++++
 drivers/char/ipmi/kcs_bmc_client.h  | 11 +++++++++++
 drivers/pci/pci.h                   |  6 ++++++
 drivers/scsi/sd.h                   |  7 +++++++
 include/linux/backing-dev-defs.h    |  6 ++++++
 include/linux/blk-mq.h              | 21 +++++++++++++++++++++
 include/linux/blk_types.h           |  6 ++++++
 include/linux/blkdev.h              | 26 ++++++++++++++++++++++++++
 include/linux/bsg-lib.h             |  6 ++++++
 include/linux/device/bus.h          |  6 ++++++
 include/linux/device/class.h        |  6 ++++++
 include/linux/fs.h                  |  6 ++++++
 include/linux/fwnode.h              |  5 +++++
 include/linux/ipmi.h                | 11 +++++++++++
 include/linux/ipmi_smi.h            | 10 ++++++++++
 include/linux/irq.h                 |  5 +++++
 include/linux/irq_work.h            |  5 +++++
 include/linux/irqdesc.h             |  5 +++++
 include/linux/irqdomain.h           |  1 +
 include/linux/mempolicy.h           |  5 +++++
 include/linux/mempool.h             |  6 ++++++
 include/linux/memremap.h            |  9 +++++++++
 include/linux/mm.h                  |  4 ++++
 include/linux/mm_types.h            |  9 +++++++--
 include/linux/net.h                 |  6 ++++++
 include/linux/nvme-fc-driver.h      |  6 ++++++
 include/linux/pagemap.h             |  5 +++++
 include/linux/pci.h                 |  5 +++++
 include/linux/sched.h               | 17 +++++++++++++++++
 include/linux/sched/topology.h      |  6 ++++++
 include/linux/shrinker.h            |  5 +++++
 include/linux/skbuff.h              |  7 ++++++-
 include/linux/swap.h                |  6 ++++++
 include/linux/timer_types.h         |  5 +++++
 include/linux/trace_events.h        | 12 +++++++++++-
 include/linux/usb/hcd.h             |  6 ++++++
 include/net/dst_ops.h               |  6 ++++++
 include/net/genetlink.h             | 10 ++++++++++
 include/net/ip6_fib.h               |  6 ++++++
 include/net/ip_fib.h                | 13 +++++++++++++
 include/net/macsec.h                |  6 ++++++
 include/net/netlink.h               |  6 ++++++
 include/net/nexthop.h               |  6 ++++++
 include/net/sch_generic.h           |  6 ++++++
 include/net/tls.h                   | 10 ++++++++++
 include/net/udp_tunnel.h            | 16 ++++++++++++++++
 include/net/xsk_buff_pool.h         |  5 +++++
 include/scsi/scsi_cmnd.h            |  6 ++++++
 include/scsi/scsi_device.h          | 13 +++++++++++++
 include/scsi/scsi_host.h            | 10 ++++++++++
 include/scsi/scsi_transport.h       |  6 ++++++
 include/scsi/scsi_transport_fc.h    |  6 ++++++
 include/scsi/scsi_transport_iscsi.h |  6 ++++++
 include/scsi/scsi_transport_sas.h   |  6 ++++++
 kernel/sched/sched.h                | 15 +++++++++++++++
 net/devlink/devl_internal.h         |  7 +++++++
 58 files changed, 452 insertions(+), 4 deletions(-)

diff --git a/block/elevator.h b/block/elevator.h
index 2a78544bf201..e3cdcae2a404 100644
--- a/block/elevator.h
+++ b/block/elevator.h
@@ -4,6 +4,7 @@
 
 #include <linux/percpu.h>
 #include <linux/hashtable.h>
+#include <linux/uek_kabi.h>
 #include "blk-mq.h"
 
 struct io_cq;
@@ -48,6 +49,11 @@ struct elevator_mq_ops {
 	struct request *(*next_request)(struct request_queue *, struct request *);
 	void (*init_icq)(struct io_cq *);
 	void (*exit_icq)(struct io_cq *);
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #define ELV_NAME_MAX	(16)
@@ -83,6 +89,11 @@ struct elevator_type
 	/* managed by elevator core */
 	char icq_cache_name[ELV_NAME_MAX + 6];	/* elvname + "_io_cq" */
 	struct list_head list;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 static inline bool elevator_tryget(struct elevator_type *e)
diff --git a/drivers/char/ipmi/ipmi_si.h b/drivers/char/ipmi/ipmi_si.h
index a7ead2a4c753..99b76445ec6e 100644
--- a/drivers/char/ipmi/ipmi_si.h
+++ b/drivers/char/ipmi/ipmi_si.h
@@ -12,6 +12,7 @@
 #include <linux/ipmi.h>
 #include <linux/interrupt.h>
 #include <linux/platform_device.h>
+#include <linux/uek_kabi.h>
 
 #define SI_DEVICE_NAME "ipmi_si"
 
@@ -66,6 +67,11 @@ struct si_sm_io {
 	u8 slave_addr;
 	enum si_type si_type;
 	struct device *dev;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 int ipmi_si_add_smi(struct si_sm_io *io);
diff --git a/drivers/char/ipmi/ipmi_si_sm.h b/drivers/char/ipmi/ipmi_si_sm.h
index c3cdbcab0f77..e1f0575c016a 100644
--- a/drivers/char/ipmi/ipmi_si_sm.h
+++ b/drivers/char/ipmi/ipmi_si_sm.h
@@ -18,6 +18,7 @@
 #define __IPMI_SI_SM_H__
 
 #include "ipmi_si.h"
+#include <linux/uek_kabi.h>
 
 /*
  * This is defined by the state machines themselves, it is an opaque
@@ -94,6 +95,11 @@ struct si_sm_handlers {
 
 	/* Return the size of the SMI structure in bytes. */
 	int (*size)(void);
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /* Current state machines that we can use. */
diff --git a/drivers/char/ipmi/kcs_bmc_client.h b/drivers/char/ipmi/kcs_bmc_client.h
index 6fdcde0a7169..e376278226d4 100644
--- a/drivers/char/ipmi/kcs_bmc_client.h
+++ b/drivers/char/ipmi/kcs_bmc_client.h
@@ -5,12 +5,18 @@
 #define __KCS_BMC_CONSUMER_H__
 
 #include <linux/irqreturn.h>
+#include <linux/uek_kabi.h>
 
 #include "kcs_bmc.h"
 
 struct kcs_bmc_driver_ops {
 	int (*add_device)(struct kcs_bmc_device *kcs_bmc);
 	int (*remove_device)(struct kcs_bmc_device *kcs_bmc);
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 struct kcs_bmc_driver {
@@ -21,6 +27,11 @@ struct kcs_bmc_driver {
 
 struct kcs_bmc_client_ops {
 	irqreturn_t (*event)(struct kcs_bmc_client *client);
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 struct kcs_bmc_client {
diff --git a/drivers/pci/pci.h b/drivers/pci/pci.h
index 8ccfaacb59f2..8e58a2170e8c 100644
--- a/drivers/pci/pci.h
+++ b/drivers/pci/pci.h
@@ -3,6 +3,7 @@
 #define DRIVERS_PCI_H
 
 #include <linux/pci.h>
+#include <linux/uek_kabi.h>
 
 /* Number of possible devfns: 0.0 to 1f.7 inclusive */
 #define MAX_NR_DEVFNS 256
@@ -405,6 +406,11 @@ struct pci_sriov {
 	u16		subsystem_device; /* VF subsystem device */
 	resource_size_t	barsz[PCI_SRIOV_NUM_BARS];	/* VF BAR size */
 	bool		drivers_autoprobe; /* Auto probing of VFs by driver */
+	
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #ifdef CONFIG_PCI_DOE
diff --git a/drivers/scsi/sd.h b/drivers/scsi/sd.h
index 36382eca941c..89b5fbc9f478 100644
--- a/drivers/scsi/sd.h
+++ b/drivers/scsi/sd.h
@@ -2,6 +2,8 @@
 #ifndef _SCSI_DISK_H
 #define _SCSI_DISK_H
 
+#include <linux/uek_kabi.h>
+
 /*
  * More than enough for everybody ;)  The huge number of majors
  * is a leftover from 16bit dev_t days, we don't really need that
@@ -156,6 +158,11 @@ struct scsi_disk {
 	unsigned	ignore_medium_access_errors : 1;
 	unsigned	rscs : 1; /* reduced stream control support */
 	unsigned	use_atomic_write_boundary : 1;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 #define to_scsi_disk(obj) container_of(obj, struct scsi_disk, disk_dev)
 
diff --git a/include/linux/backing-dev-defs.h b/include/linux/backing-dev-defs.h
index 2ad261082bba..b11a3edc5afb 100644
--- a/include/linux/backing-dev-defs.h
+++ b/include/linux/backing-dev-defs.h
@@ -13,6 +13,7 @@
 #include <linux/workqueue.h>
 #include <linux/kref.h>
 #include <linux/refcount.h>
+#include <linux/uek_kabi.h>
 
 struct page;
 struct device;
@@ -158,6 +159,11 @@ struct bdi_writeback {
 		struct rcu_head rcu;
 	};
 #endif
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 struct backing_dev_info {
diff --git a/include/linux/blk-mq.h b/include/linux/blk-mq.h
index 7b5e5388c380..921d562a89e9 100644
--- a/include/linux/blk-mq.h
+++ b/include/linux/blk-mq.h
@@ -9,6 +9,7 @@
 #include <linux/prefetch.h>
 #include <linux/srcu.h>
 #include <linux/rw_hint.h>
+#include <linux/uek_kabi.h>
 
 struct blk_mq_tags;
 struct blk_flush_queue;
@@ -445,6 +446,11 @@ struct blk_mq_hw_ctx {
 	struct dentry		*sched_debugfs_dir;
 #endif
 
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	/**
 	 * @hctx_list: if this hctx is not in use, this is an entry in
 	 * q->unused_hctx_list.
@@ -536,6 +542,11 @@ struct blk_mq_tag_set {
 	struct mutex		tag_list_lock;
 	struct list_head	tag_list;
 	struct srcu_struct	*srcu;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /**
@@ -547,6 +558,11 @@ struct blk_mq_tag_set {
 struct blk_mq_queue_data {
 	struct request *rq;
 	bool last;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 typedef bool (busy_tag_iter_fn)(struct request *, void *);
@@ -666,6 +682,11 @@ struct blk_mq_ops {
 	 */
 	void (*show_rq)(struct seq_file *m, struct request *rq);
 #endif
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /* Keep hctx_flag_name[] in sync with the definitions below */
diff --git a/include/linux/blk_types.h b/include/linux/blk_types.h
index 0874218454d7..be1d20199168 100644
--- a/include/linux/blk_types.h
+++ b/include/linux/blk_types.h
@@ -11,6 +11,7 @@
 #include <linux/device.h>
 #include <linux/ktime.h>
 #include <linux/rw_hint.h>
+#include <linux/uek_kabi.h>
 
 struct bio_set;
 struct bio;
@@ -270,6 +271,11 @@ struct bio {
 
 	struct bio_set		*bi_pool;
 
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	/*
 	 * We can inline a number of vecs at the end of the bio, to avoid
 	 * double allocations for a small number of bio_vecs. This member
diff --git a/include/linux/blkdev.h b/include/linux/blkdev.h
index b8471a49fd1a..7f692c4878a0 100644
--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -26,6 +26,7 @@
 #include <linux/xarray.h>
 #include <linux/file.h>
 #include <linux/lockdep.h>
+#include <linux/uek_kabi.h>
 
 struct module;
 struct request_queue;
@@ -120,6 +121,11 @@ struct blk_integrity {
 	unsigned char				pi_offset;
 	unsigned char				interval_exp;
 	unsigned char				tag_size;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 typedef unsigned int __bitwise blk_mode_t;
@@ -218,6 +224,11 @@ struct gendisk {
 	 * devices that do not have multiple independent access ranges.
 	 */
 	struct blk_independent_access_ranges *ia_ranges;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /**
@@ -406,6 +417,11 @@ struct queue_limits {
 	unsigned int		dma_pad_mask;
 
 	struct blk_integrity	integrity;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 typedef int (*report_zones_cb)(struct blk_zone *zone, unsigned int idx,
@@ -598,6 +614,11 @@ struct request_queue {
 	struct mutex		debugfs_mutex;
 
 	bool			mq_sysfs_init_done;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /* Keep blk_queue_flag_name[] in sync with the definitions below */
@@ -1572,6 +1593,11 @@ struct block_device_operations {
 	 * driver.
 	 */
 	int (*alternative_gpt_sector)(struct gendisk *disk, sector_t *sector);
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #ifdef CONFIG_COMPAT
diff --git a/include/linux/bsg-lib.h b/include/linux/bsg-lib.h
index 14fa93268630..3d4d2ccc6258 100644
--- a/include/linux/bsg-lib.h
+++ b/include/linux/bsg-lib.h
@@ -10,6 +10,7 @@
 #define _BLK_BSG_
 
 #include <linux/blkdev.h>
+#include <linux/uek_kabi.h>
 
 struct bsg_job;
 struct request;
@@ -60,6 +61,11 @@ struct bsg_job {
 	struct bio *bidi_bio;
 
 	void *dd_data;		/* Used for driver-specific storage */
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 void bsg_job_done(struct bsg_job *job, int result,
diff --git a/include/linux/device/bus.h b/include/linux/device/bus.h
index cdc4757217f9..88b4b409116a 100644
--- a/include/linux/device/bus.h
+++ b/include/linux/device/bus.h
@@ -17,6 +17,7 @@
 #include <linux/kobject.h>
 #include <linux/klist.h>
 #include <linux/pm.h>
+#include <linux/uek_kabi.h>
 
 struct device_driver;
 struct fwnode_handle;
@@ -102,6 +103,11 @@ struct bus_type {
 	const struct dev_pm_ops *pm;
 
 	bool need_parent_lock;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 int __must_check bus_register(const struct bus_type *bus);
diff --git a/include/linux/device/class.h b/include/linux/device/class.h
index 518c9c83d64b..0b0c0a5e05c5 100644
--- a/include/linux/device/class.h
+++ b/include/linux/device/class.h
@@ -18,6 +18,7 @@
 #include <linux/klist.h>
 #include <linux/pm.h>
 #include <linux/device/bus.h>
+#include <linux/uek_kabi.h>
 
 struct device;
 struct fwnode_handle;
@@ -67,6 +68,11 @@ struct class {
 	void (*get_ownership)(const struct device *dev, kuid_t *uid, kgid_t *gid);
 
 	const struct dev_pm_ops *pm;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 struct class_dev_iter {
diff --git a/include/linux/fs.h b/include/linux/fs.h
index 7989a7bfe346..8cf03da44a06 100644
--- a/include/linux/fs.h
+++ b/include/linux/fs.h
@@ -45,6 +45,7 @@
 #include <linux/slab.h>
 #include <linux/maple_tree.h>
 #include <linux/rw_hint.h>
+#include <linux/uek_kabi.h>
 
 #include <asm/byteorder.h>
 #include <uapi/linux/fs.h>
@@ -483,6 +484,11 @@ struct address_space {
 	struct list_head	i_private_list;
 	struct rw_semaphore	i_mmap_rwsem;
 	void *			i_private_data;
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+	
 } __attribute__((aligned(sizeof(long)))) __randomize_layout;
 	/*
 	 * On most architectures that alignment is already the case; but
diff --git a/include/linux/fwnode.h b/include/linux/fwnode.h
index c7502d338023..3137731456da 100644
--- a/include/linux/fwnode.h
+++ b/include/linux/fwnode.h
@@ -13,6 +13,7 @@
 #include <linux/err.h>
 #include <linux/list.h>
 #include <linux/types.h>
+#include <linux/uek_kabi.h>
 
 enum dev_dma_attr {
 	DEV_DMA_NOT_SUPPORTED,
@@ -53,6 +54,10 @@ struct fwnode_handle {
 	struct list_head suppliers;
 	struct list_head consumers;
 	u8 flags;
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /*
diff --git a/include/linux/ipmi.h b/include/linux/ipmi.h
index a1c9c0d48ebf..b31d5fdcde8a 100644
--- a/include/linux/ipmi.h
+++ b/include/linux/ipmi.h
@@ -19,6 +19,7 @@
 #include <linux/list.h>
 #include <linux/proc_fs.h>
 #include <linux/acpi.h> /* For acpi_handle */
+#include <linux/uek_kabi.h>
 
 struct module;
 struct device;
@@ -70,6 +71,11 @@ struct ipmi_recv_msg {
 	 * the size or existence of this, since it may change.
 	 */
 	unsigned char   msg_data[IPMI_MAX_MSG_LENGTH];
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #define INIT_IPMI_RECV_MSG(done_handler) \
@@ -110,6 +116,11 @@ struct ipmi_user_hndl {
 	 * if it is not usable.
 	 */
 	void (*shutdown)(void *handler_data);
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /* Create a new user of the IPMI layer on the given interface number. */
diff --git a/include/linux/ipmi_smi.h b/include/linux/ipmi_smi.h
index 5d69820d8b02..e0f7eabb10cb 100644
--- a/include/linux/ipmi_smi.h
+++ b/include/linux/ipmi_smi.h
@@ -19,6 +19,7 @@
 #include <linux/proc_fs.h>
 #include <linux/platform_device.h>
 #include <linux/ipmi.h>
+#include <linux/uek_kabi.h>
 
 struct device;
 
@@ -123,6 +124,10 @@ struct ipmi_smi_msg {
 	 * (presumably to free it).
 	 */
 	void (*done)(struct ipmi_smi_msg *msg);
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #define INIT_IPMI_SMI_MSG(done_handler) \
@@ -218,6 +223,11 @@ struct ipmi_smi_handlers {
 	 * block.
 	 */
 	void (*set_maintenance_mode)(void *send_info, bool enable);
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 struct ipmi_device_id {
diff --git a/include/linux/irq.h b/include/linux/irq.h
index fa711f80957b..e91a02a63a25 100644
--- a/include/linux/irq.h
+++ b/include/linux/irq.h
@@ -23,6 +23,7 @@
 #include <asm/irq.h>
 #include <asm/ptrace.h>
 #include <asm/irq_regs.h>
+#include <linux/uek_kabi.h>
 
 struct seq_file;
 struct module;
@@ -187,6 +188,10 @@ struct irq_data {
 	struct irq_data		*parent_data;
 #endif
 	void			*chip_data;
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /*
diff --git a/include/linux/irq_work.h b/include/linux/irq_work.h
index 136f2980cba3..34dc5d0e0c5a 100644
--- a/include/linux/irq_work.h
+++ b/include/linux/irq_work.h
@@ -4,6 +4,7 @@
 
 #include <linux/smp_types.h>
 #include <linux/rcuwait.h>
+#include <linux/uek_kabi.h>
 
 /*
  * An entry can be in one of four states:
@@ -18,6 +19,10 @@ struct irq_work {
 	struct __call_single_node node;
 	void (*func)(struct irq_work *);
 	struct rcuwait irqwait;
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #define __IRQ_WORK_INIT(_func, _flags) (struct irq_work){	\
diff --git a/include/linux/irqdesc.h b/include/linux/irqdesc.h
index fd091c35d572..8538b0a80118 100644
--- a/include/linux/irqdesc.h
+++ b/include/linux/irqdesc.h
@@ -5,6 +5,7 @@
 #include <linux/rcupdate.h>
 #include <linux/kobject.h>
 #include <linux/mutex.h>
+#include <linux/uek_kabi.h>
 
 /*
  * Core internal functions to deal with irq descriptors
@@ -117,6 +118,10 @@ struct irq_desc {
 #ifdef CONFIG_HARDIRQS_SW_RESEND
 	struct hlist_node	resend_node;
 #endif
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 } ____cacheline_internodealigned_in_smp;
 
 #ifdef CONFIG_SPARSE_IRQ
diff --git a/include/linux/irqdomain.h b/include/linux/irqdomain.h
index b3213a966256..c408560ac0a1 100644
--- a/include/linux/irqdomain.h
+++ b/include/linux/irqdomain.h
@@ -36,6 +36,7 @@
 #include <linux/of.h>
 #include <linux/mutex.h>
 #include <linux/radix-tree.h>
+#include <linux/uek_kabi.h>
 
 struct device_node;
 struct fwnode_handle;
diff --git a/include/linux/mempolicy.h b/include/linux/mempolicy.h
index 1add16f21612..d452c4ecb967 100644
--- a/include/linux/mempolicy.h
+++ b/include/linux/mempolicy.h
@@ -14,6 +14,7 @@
 #include <linux/nodemask.h>
 #include <linux/pagemap.h>
 #include <uapi/linux/mempolicy.h>
+#include <linux/uek_kabi.h>
 
 struct mm_struct;
 
@@ -54,6 +55,10 @@ struct mempolicy {
 		nodemask_t cpuset_mems_allowed;	/* relative to these nodes */
 		nodemask_t user_nodemask;	/* nodemask passed by user */
 	} w;
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /*
diff --git a/include/linux/mempool.h b/include/linux/mempool.h
index 7b151441341b..be7717ff5159 100644
--- a/include/linux/mempool.h
+++ b/include/linux/mempool.h
@@ -9,6 +9,7 @@
 #include <linux/alloc_tag.h>
 #include <linux/wait.h>
 #include <linux/compiler.h>
+#include <linux/uek_kabi.h>
 
 struct kmem_cache;
 
@@ -25,6 +26,11 @@ typedef struct mempool_s {
 	mempool_alloc_t *alloc;
 	mempool_free_t *free;
 	wait_queue_head_t wait;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 } mempool_t;
 
 static inline bool mempool_initialized(mempool_t *pool)
diff --git a/include/linux/memremap.h b/include/linux/memremap.h
index 4d17053f6320..16eca89d1be3 100644
--- a/include/linux/memremap.h
+++ b/include/linux/memremap.h
@@ -7,6 +7,7 @@
 #include <linux/range.h>
 #include <linux/ioport.h>
 #include <linux/percpu-refcount.h>
+#include <linux/uek_kabi.h>
 
 struct resource;
 struct device;
@@ -27,6 +28,10 @@ struct vmem_altmap {
 	unsigned long align;
 	unsigned long alloc;
 	bool inaccessible;
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /*
@@ -101,6 +106,10 @@ struct dev_pagemap_ops {
 	 */
 	int (*memory_failure)(struct dev_pagemap *pgmap, unsigned long pfn,
 			      unsigned long nr_pages, int mf_flags);
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #define PGMAP_ALTMAP_VALID	(1 << 0)
diff --git a/include/linux/mm.h b/include/linux/mm.h
index 337004a4db10..4b369fcc1d43 100644
--- a/include/linux/mm.h
+++ b/include/linux/mm.h
@@ -602,6 +602,10 @@ struct vm_fault {
 					 * page table to avoid allocation from
 					 * atomic context.
 					 */
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)		
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /*
diff --git a/include/linux/mm_types.h b/include/linux/mm_types.h
index cdc651b6992f..7627a9892e87 100644
--- a/include/linux/mm_types.h
+++ b/include/linux/mm_types.h
@@ -1085,16 +1085,21 @@ struct mm_struct {
 #endif
 		} lru_gen;
 #endif /* CONFIG_LRU_GEN_WALKS_MMU */
-	} __randomize_layout;
-
 	UEK_KABI_RESERVE(1)
 	UEK_KABI_RESERVE(2)
 	UEK_KABI_RESERVE(3)
 	UEK_KABI_RESERVE(4)
+
+	} __randomize_layout;
+
 	UEK_KABI_RESERVE(5)
 	UEK_KABI_RESERVE(6)
 	UEK_KABI_RESERVE(7)
 	UEK_KABI_RESERVE(8)
+	UEK_KABI_RESERVE(9)
+	UEK_KABI_RESERVE(10)
+	UEK_KABI_RESERVE(11)
+	UEK_KABI_RESERVE(12)
 
 	/*
 	 * The mm_cpumask needs to be at the end of mm_struct, because it
diff --git a/include/linux/net.h b/include/linux/net.h
index 94660492e211..b11cbc1e4b1e 100644
--- a/include/linux/net.h
+++ b/include/linux/net.h
@@ -24,6 +24,7 @@
 #include <linux/fs.h>
 #include <linux/mm.h>
 #include <linux/sockptr.h>
+#include <linux/uek_kabi.h>
 
 #include <uapi/linux/net.h>
 
@@ -103,6 +104,11 @@ struct socket_wq {
 	struct fasync_struct	*fasync_list;
 	unsigned long		flags; /* %SOCKWQ_ASYNC_NOSPACE, etc */
 	struct rcu_head		rcu;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 } ____cacheline_aligned_in_smp;
 
 /**
diff --git a/include/linux/nvme-fc-driver.h b/include/linux/nvme-fc-driver.h
index 9f6acadfe0c8..46ead29b8c0e 100644
--- a/include/linux/nvme-fc-driver.h
+++ b/include/linux/nvme-fc-driver.h
@@ -8,6 +8,7 @@
 
 #include <linux/scatterlist.h>
 #include <linux/blk-mq.h>
+#include <linux/uek_kabi.h>
 
 
 /*
@@ -511,6 +512,11 @@ struct nvme_fc_port_template {
 	u32	remote_priv_sz;
 	u32	lsrqst_priv_sz;
 	u32	fcprqst_priv_sz;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 
diff --git a/include/linux/pagemap.h b/include/linux/pagemap.h
index a72e04d87191..e083dd1400b1 100644
--- a/include/linux/pagemap.h
+++ b/include/linux/pagemap.h
@@ -15,6 +15,7 @@
 #include <linux/bitops.h>
 #include <linux/hardirq.h> /* for in_interrupt() */
 #include <linux/hugetlb_inline.h>
+#include <linux/uek_kabi.h>
 
 struct folio_batch;
 
@@ -1369,6 +1370,10 @@ struct readahead_control {
 	struct file *file;
 	struct address_space *mapping;
 	struct file_ra_state *ra;
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 /* private: use the readahead_* accessors instead */
 	pgoff_t _index;
 	unsigned int _nr_pages;
diff --git a/include/linux/pci.h b/include/linux/pci.h
index 9c3bb42d582f..a15ba0f108c8 100644
--- a/include/linux/pci.h
+++ b/include/linux/pci.h
@@ -40,6 +40,7 @@
 #include <linux/resource_ext.h>
 #include <linux/msi_api.h>
 #include <uapi/linux/pci.h>
+#include <linux/uek_kabi.h>
 
 #include <linux/pci_ids.h>
 
@@ -986,6 +987,10 @@ struct pci_driver {
 	struct device_driver	driver;
 	struct pci_dynids	dynids;
 	bool driver_managed_dma;
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #define to_pci_driver(__drv)	\
diff --git a/include/linux/sched.h b/include/linux/sched.h
index 2ca95504ef41..68bff207f318 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -488,6 +488,10 @@ struct sched_avg {
 	unsigned long			runnable_avg;
 	unsigned long			util_avg;
 	unsigned int			util_est;
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 } ____cacheline_aligned;
 
 /*
@@ -613,6 +617,11 @@ struct sched_rt_entity {
 	/* rq "owned" by this entity/group: */
 	struct rt_rq			*my_q;
 #endif
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 } __randomize_layout;
 
 typedef bool (*dl_server_has_tasks_f)(struct sched_dl_entity *);
@@ -761,6 +770,10 @@ struct uclamp_se {
 	unsigned int bucket_id		: bits_per(UCLAMP_BUCKETS);
 	unsigned int active		: 1;
 	unsigned int user_defined	: 1;
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 #endif /* CONFIG_UCLAMP_TASK */
 
@@ -1635,6 +1648,10 @@ struct task_struct {
 	struct user_event_mm		*user_event_mm;
 #endif
 
+	UEK_KABI_RESERVE(11)
+	UEK_KABI_RESERVE(12)
+	UEK_KABI_RESERVE(13)
+	UEK_KABI_RESERVE(14)
 	/*
 	 * New fields for task_struct should be added above here, so that
 	 * they are included in the randomized portion of task_struct.
diff --git a/include/linux/sched/topology.h b/include/linux/sched/topology.h
index debd549830cd..a6a7ea3ef55d 100644
--- a/include/linux/sched/topology.h
+++ b/include/linux/sched/topology.h
@@ -3,6 +3,7 @@
 #define _LINUX_SCHED_TOPOLOGY_H
 
 #include <linux/topology.h>
+#include <linux/uek_kabi.h>
 
 #include <linux/sched/idle.h>
 
@@ -150,6 +151,11 @@ struct sched_domain {
 	};
 	struct sched_domain_shared *shared;
 
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	unsigned int span_weight;
 	/*
 	 * Span of all CPUs in this domain.
diff --git a/include/linux/shrinker.h b/include/linux/shrinker.h
index 1a00be90d93a..19a5d9e08223 100644
--- a/include/linux/shrinker.h
+++ b/include/linux/shrinker.h
@@ -6,6 +6,7 @@
 #include <linux/types.h>
 #include <linux/refcount.h>
 #include <linux/completion.h>
+#include <linux/uek_kabi.h>
 
 #define SHRINKER_UNIT_BITS	BITS_PER_LONG
 
@@ -115,6 +116,10 @@ struct shrinker {
 #endif
 	/* objs pending delete, per node */
 	atomic_long_t *nr_deferred;
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 #define DEFAULT_SEEKS 2 /* A good number if you don't know better. */
 
diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
index 8dad9bb5de4f..248d22edb80c 100644
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -1072,6 +1072,11 @@ struct sk_buff {
 	u64			kcov_handle;
 #endif
 
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	); /* end headers group */
 
 	/* These elements must be at the end, see alloc_skb() for details.  */
@@ -1086,7 +1091,7 @@ struct sk_buff {
 	/* only usable after checking ->active_extensions != 0 */
 	struct skb_ext		*extensions;
 #endif
-	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(5)
 };
 
 /* if you move pkt_type around you also must adapt those constants */
diff --git a/include/linux/swap.h b/include/linux/swap.h
index 3f42b436e74e..a0706ae44bd6 100644
--- a/include/linux/swap.h
+++ b/include/linux/swap.h
@@ -350,6 +350,12 @@ struct swap_info_struct {
 	struct work_struct discard_work; /* discard worker */
 	struct work_struct reclaim_work; /* reclaim worker */
 	struct list_head discard_clusters; /* discard clusters list */
+
+	UEK_KABI_RESERVE(9)
+	UEK_KABI_RESERVE(10)
+	UEK_KABI_RESERVE(11)
+	UEK_KABI_RESERVE(12)
+	
 	struct plist_node avail_lists[]; /*
 					   * entries in swap_avail_heads, one
 					   * entry per node.
diff --git a/include/linux/timer_types.h b/include/linux/timer_types.h
index fae5a388f914..7d2d71012ad3 100644
--- a/include/linux/timer_types.h
+++ b/include/linux/timer_types.h
@@ -4,6 +4,7 @@
 
 #include <linux/lockdep_types.h>
 #include <linux/types.h>
+#include <linux/uek_kabi.h>
 
 struct timer_list {
 	/*
@@ -18,6 +19,10 @@ struct timer_list {
 #ifdef CONFIG_LOCKDEP
 	struct lockdep_map	lockdep_map;
 #endif
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #endif /* _LINUX_TIMER_TYPES_H */
diff --git a/include/linux/trace_events.h b/include/linux/trace_events.h
index 3eceb5c08a80..4c6383c983b5 100644
--- a/include/linux/trace_events.h
+++ b/include/linux/trace_events.h
@@ -9,6 +9,7 @@
 #include <linux/hardirq.h>
 #include <linux/perf_event.h>
 #include <linux/tracepoint.h>
+#include <linux/uek_kabi.h>
 
 struct trace_array;
 struct array_buffer;
@@ -119,6 +120,11 @@ struct trace_iterator {
 	/* it's true when current open file is snapshot */
 	bool			snapshot;
 
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	/* The below is zeroed out in pipe_read */
 	struct trace_seq	seq;
 	struct trace_entry	*ent;
@@ -130,7 +136,7 @@ struct trace_iterator {
 
 	loff_t			pos;
 	long			idx;
-	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(5)
 
 	/* All new field here will be zeroed out in pipe_read */
 };
@@ -688,6 +694,10 @@ struct trace_event_file {
 	refcount_t		ref;	/* ref count for opened files */
 	atomic_t		sm_ref;	/* soft-mode reference counter */
 	atomic_t		tm_ref;	/* trigger-mode reference counter */
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #ifdef CONFIG_HIST_TRIGGERS
diff --git a/include/linux/usb/hcd.h b/include/linux/usb/hcd.h
index ac95e7c89df5..f6998dae7e38 100644
--- a/include/linux/usb/hcd.h
+++ b/include/linux/usb/hcd.h
@@ -11,6 +11,7 @@
 #include <linux/rwsem.h>
 #include <linux/interrupt.h>
 #include <linux/idr.h>
+#include <linux/uek_kabi.h>
 
 #define MAX_TOPO_LEVEL		6
 
@@ -213,6 +214,11 @@ struct usb_hcd {
 	 * (ohci 32, uhci 1024, ehci 256/512/1024).
 	 */
 
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	/* The HC driver's private data is stored at the end of
 	 * this structure.
 	 */
diff --git a/include/net/dst_ops.h b/include/net/dst_ops.h
index 836e705d7b79..9e4170f0e47a 100644
--- a/include/net/dst_ops.h
+++ b/include/net/dst_ops.h
@@ -51,6 +51,12 @@ struct dst_ops {
 	UEK_KABI_RESERVE(8)
 
 	struct percpu_counter	pcpuc_entries ____cacheline_aligned_in_smp;
+
+	UEK_KABI_RESERVE(9)
+	UEK_KABI_RESERVE(10)
+	UEK_KABI_RESERVE(11)
+	UEK_KABI_RESERVE(12)
+
 };
 
 static inline int dst_entries_get_fast(struct dst_ops *dst)
diff --git a/include/net/genetlink.h b/include/net/genetlink.h
index fe7d309f3d22..7f130ec75c84 100644
--- a/include/net/genetlink.h
+++ b/include/net/genetlink.h
@@ -201,6 +201,11 @@ struct genl_small_ops {
 	u8	internal_flags;
 	u8	flags;
 	u8	validate;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /**
@@ -289,6 +294,11 @@ struct genl_split_ops {
 	u8			internal_flags;
 	u8			flags;
 	u8			validate;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /**
diff --git a/include/net/ip6_fib.h b/include/net/ip6_fib.h
index 8d4ef809cacc..87e6682f67d2 100644
--- a/include/net/ip6_fib.h
+++ b/include/net/ip6_fib.h
@@ -201,6 +201,12 @@ struct fib6_info {
 
 	struct rcu_head			rcu;
 	struct nexthop			*nh;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	struct fib6_nh			fib6_nh[];
 };
 
diff --git a/include/net/ip_fib.h b/include/net/ip_fib.h
index 967e4dc555fa..c00e48768a95 100644
--- a/include/net/ip_fib.h
+++ b/include/net/ip_fib.h
@@ -24,6 +24,7 @@
 #include <linux/refcount.h>
 #include <linux/ip.h>
 #include <linux/in_route.h>
+#include <linux/uek_kabi.h>
 
 struct fib_config {
 	u8			fc_dst_len;
@@ -159,6 +160,12 @@ struct fib_info {
 	bool			pfsrc_removed;
 	struct nexthop		*nh;
 	struct rcu_head		rcu;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+	
 	struct fib_nh		fib_nh[] __counted_by(fib_nhs);
 };
 
@@ -258,6 +265,12 @@ struct fib_table {
 	int			tb_num_default;
 	struct rcu_head		rcu;
 	unsigned long 		*tb_data;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	unsigned long		__data[];
 };
 
diff --git a/include/net/macsec.h b/include/net/macsec.h
index de216cbc6b05..4832e93f715c 100644
--- a/include/net/macsec.h
+++ b/include/net/macsec.h
@@ -11,6 +11,7 @@
 #include <linux/if_vlan.h>
 #include <uapi/linux/if_link.h>
 #include <uapi/linux/if_macsec.h>
+#include <linux/uek_kabi.h>
 
 #define MACSEC_DEFAULT_PN_LEN 4
 #define MACSEC_XPN_PN_LEN 8
@@ -354,6 +355,11 @@ struct macsec_ops {
 	unsigned int needed_headroom;
 	unsigned int needed_tailroom;
 	bool rx_uses_md_dst;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 void macsec_pn_wrapped(struct macsec_secy *secy, struct macsec_tx_sa *tx_sa);
diff --git a/include/net/netlink.h b/include/net/netlink.h
index db6af207287c..64af6250af31 100644
--- a/include/net/netlink.h
+++ b/include/net/netlink.h
@@ -6,6 +6,7 @@
 #include <linux/netlink.h>
 #include <linux/jiffies.h>
 #include <linux/in6.h>
+#include <linux/uek_kabi.h>
 
 /* ========================================================================
  *         Netlink Messages and Attributes Interface (As Seen On TV)
@@ -373,6 +374,11 @@ struct nla_policy {
 		int (*validate)(const struct nlattr *attr,
 				struct netlink_ext_ack *extack);
 	};
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #define NLA_POLICY_ETH_ADDR		NLA_POLICY_EXACT_LEN(ETH_ALEN)
diff --git a/include/net/nexthop.h b/include/net/nexthop.h
index d9fb44e8b321..547e0b529f90 100644
--- a/include/net/nexthop.h
+++ b/include/net/nexthop.h
@@ -16,6 +16,7 @@
 #include <net/ip_fib.h>
 #include <net/ip6_fib.h>
 #include <net/netlink.h>
+#include <linux/uek_kabi.h>
 
 #define NEXTHOP_VALID_USER_FLAGS RTNH_F_ONLINK
 
@@ -135,6 +136,11 @@ struct nh_group {
 	bool			has_v4;
 	bool			hw_stats;
 
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	struct nh_res_table __rcu *res_table;
 	struct nh_grp_entry	nh_entries[] __counted_by(num_nh);
 };
diff --git a/include/net/sch_generic.h b/include/net/sch_generic.h
index 2c002c0836cf..0663fd6fc02b 100644
--- a/include/net/sch_generic.h
+++ b/include/net/sch_generic.h
@@ -130,6 +130,12 @@ struct Qdisc {
 	struct rcu_head		rcu;
 	netdevice_tracker	dev_tracker;
 	struct lock_class_key	root_lock_key;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	/* private data */
 	long privdata[] ____cacheline_aligned;
 };
diff --git a/include/net/tls.h b/include/net/tls.h
index e618ea561cb5..9349f87fcefe 100644
--- a/include/net/tls.h
+++ b/include/net/tls.h
@@ -205,6 +205,11 @@ union tls_crypto_context {
 		struct tls12_crypto_info_sm4_gcm sm4_gcm;
 		struct tls12_crypto_info_sm4_ccm sm4_ccm;
 	};
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 struct tls_prot_info {
@@ -282,6 +287,11 @@ struct tlsdev_ops {
 	int (*tls_dev_resync)(struct net_device *netdev,
 			      struct sock *sk, u32 seq, u8 *rcd_sn,
 			      enum tls_offload_ctx_dir direction);
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 enum tls_offload_sync_type {
diff --git a/include/net/udp_tunnel.h b/include/net/udp_tunnel.h
index 2787b4eec8eb..2c2f19d781bb 100644
--- a/include/net/udp_tunnel.h
+++ b/include/net/udp_tunnel.h
@@ -4,6 +4,7 @@
 
 #include <net/ip_tunnels.h>
 #include <net/udp.h>
+#include <linux/uek_kabi.h>
 
 #if IS_ENABLED(CONFIG_IPV6)
 #include <net/ipv6.h>
@@ -231,6 +232,11 @@ struct udp_tunnel_nic_shared {
 	struct udp_tunnel_nic *udp_tunnel_nic_info;
 
 	struct list_head devices;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 struct udp_tunnel_nic_shared_node {
@@ -286,7 +292,17 @@ struct udp_tunnel_nic_info {
 	struct udp_tunnel_nic_table_info {
 		unsigned int n_entries;
 		unsigned int tunnel_types;
+
+		UEK_KABI_RESERVE(1)
+		UEK_KABI_RESERVE(2)
+		UEK_KABI_RESERVE(3)
+		UEK_KABI_RESERVE(4)
 	} tables[UDP_TUNNEL_NIC_MAX_TABLES];
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /* UDP tunnel module dependencies
diff --git a/include/net/xsk_buff_pool.h b/include/net/xsk_buff_pool.h
index 823fd5c7a3b1..3b9c7e58fd03 100644
--- a/include/net/xsk_buff_pool.h
+++ b/include/net/xsk_buff_pool.h
@@ -9,6 +9,7 @@
 #include <linux/dma-mapping.h>
 #include <linux/bpf.h>
 #include <net/xdp.h>
+#include <linux/uek_kabi.h>
 
 struct xsk_buff_pool;
 struct xdp_rxq_info;
@@ -91,6 +92,10 @@ struct xsk_buff_pool {
 	 * sockets share a single cq when the same netdev and queue id is shared.
 	 */
 	spinlock_t cq_lock;
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 	struct xdp_buff_xsk *free_heads[];
 };
 
diff --git a/include/scsi/scsi_cmnd.h b/include/scsi/scsi_cmnd.h
index 8ecfb94049db..a00eca732cd2 100644
--- a/include/scsi/scsi_cmnd.h
+++ b/include/scsi/scsi_cmnd.h
@@ -10,6 +10,7 @@
 #include <linux/timer.h>
 #include <linux/scatterlist.h>
 #include <scsi/scsi_device.h>
+#include <linux/uek_kabi.h>
 
 struct Scsi_Host;
 
@@ -141,6 +142,11 @@ struct scsi_cmnd {
 					 * to be at an address < 16Mb). */
 
 	int result;		/* Status code from lower level driver */
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /* Variant of blk_mq_rq_from_pdu() that verifies the type of its argument. */
diff --git a/include/scsi/scsi_device.h b/include/scsi/scsi_device.h
index 9c540f5468eb..251d8985c0af 100644
--- a/include/scsi/scsi_device.h
+++ b/include/scsi/scsi_device.h
@@ -9,6 +9,7 @@
 #include <scsi/scsi.h>
 #include <linux/atomic.h>
 #include <linux/sbitmap.h>
+#include <linux/uek_kabi.h>
 
 struct bsg_device;
 struct device;
@@ -281,6 +282,12 @@ struct scsi_device {
 	struct mutex		state_mutex;
 	enum scsi_device_state sdev_state;
 	struct task_struct	*quiesced_by;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	unsigned long		sdev_data[];
 } __attribute__((aligned(sizeof(unsigned long))));
 
@@ -367,6 +374,12 @@ struct scsi_target {
 	char			scsi_level;
 	enum scsi_target_state	state;
 	void 			*hostdata; /* available to low-level driver */
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	unsigned long		starget_data[]; /* for the transport */
 	/* starget_data must be the last element!!!! */
 } __attribute__((aligned(sizeof(unsigned long))));
diff --git a/include/scsi/scsi_host.h b/include/scsi/scsi_host.h
index 2b4ab0369ffb..0716a7aaaf57 100644
--- a/include/scsi/scsi_host.h
+++ b/include/scsi/scsi_host.h
@@ -10,6 +10,7 @@
 #include <linux/seq_file.h>
 #include <linux/blk-mq.h>
 #include <scsi/scsi.h>
+#include <linux/uek_kabi.h>
 
 struct block_device;
 struct completion;
@@ -503,6 +504,11 @@ struct scsi_host_template {
 	 *   scsi_netlink.h
 	 */
 	u64 vendor_id;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /*
@@ -720,6 +726,10 @@ struct Scsi_Host {
 
 	/* Delay for runtime autosuspend */
 	int rpm_autosuspend_delay;
+	UEK_KABI_RESERVE(5)
+	UEK_KABI_RESERVE(6)
+	UEK_KABI_RESERVE(7)
+	UEK_KABI_RESERVE(8)
 
 	/*
 	 * We should ensure that this is aligned, both for better performance
diff --git a/include/scsi/scsi_transport.h b/include/scsi/scsi_transport.h
index 1394cf313bb3..fd2772fef305 100644
--- a/include/scsi/scsi_transport.h
+++ b/include/scsi/scsi_transport.h
@@ -12,6 +12,7 @@
 #include <linux/bug.h>
 #include <scsi/scsi_host.h>
 #include <scsi/scsi_device.h>
+#include <linux/uek_kabi.h>
 
 struct scsi_transport_template {
 	/* the attribute containers */
@@ -43,6 +44,11 @@ struct scsi_transport_template {
 	 * Allows a transport to override the default error handler.
 	 */
 	void (* eh_strategy_handler)(struct Scsi_Host *);
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #define transport_class_to_shost(tc) \
diff --git a/include/scsi/scsi_transport_fc.h b/include/scsi/scsi_transport_fc.h
index d02b55261307..c6bd8804734d 100644
--- a/include/scsi/scsi_transport_fc.h
+++ b/include/scsi/scsi_transport_fc.h
@@ -16,6 +16,7 @@
 #include <scsi/scsi.h>
 #include <scsi/scsi_netlink.h>
 #include <scsi/scsi_host.h>
+#include <linux/uek_kabi.h>
 
 struct scsi_transport_template;
 
@@ -760,6 +761,11 @@ struct fc_function_template {
 	unsigned long	show_host_system_hostname:1;
 
 	unsigned long	disable_target_scan:1;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /**
diff --git a/include/scsi/scsi_transport_iscsi.h b/include/scsi/scsi_transport_iscsi.h
index bd1243657c01..bfe12499f876 100644
--- a/include/scsi/scsi_transport_iscsi.h
+++ b/include/scsi/scsi_transport_iscsi.h
@@ -14,6 +14,7 @@
 #include <linux/list.h>
 #include <linux/mutex.h>
 #include <scsi/iscsi_if.h>
+#include <linux/uek_kabi.h>
 
 struct scsi_transport_template;
 struct iscsi_transport;
@@ -156,6 +157,11 @@ struct iscsi_transport {
 	int (*logout_flashnode_sid) (struct iscsi_cls_session *cls_sess);
 	int (*get_host_stats) (struct Scsi_Host *shost, char *buf, int len);
 	u8 (*check_protection)(struct iscsi_task *task, sector_t *sector);
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 /*
diff --git a/include/scsi/scsi_transport_sas.h b/include/scsi/scsi_transport_sas.h
index e3b6ce3cbf88..0015ec8c445e 100644
--- a/include/scsi/scsi_transport_sas.h
+++ b/include/scsi/scsi_transport_sas.h
@@ -7,6 +7,7 @@
 #include <linux/mutex.h>
 #include <scsi/sas.h>
 #include <linux/bsg-lib.h>
+#include <linux/uek_kabi.h>
 
 struct scsi_transport_template;
 struct sas_rphy;
@@ -182,6 +183,11 @@ struct sas_function_template {
 	int (*set_phy_speed)(struct sas_phy *, struct sas_phy_linkrates *);
 	void (*smp_handler)(struct bsg_job *, struct Scsi_Host *,
 			struct sas_rphy *);
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 
diff --git a/kernel/sched/sched.h b/kernel/sched/sched.h
index f1ff9a17c5fc..8da5cf5d7f6c 100644
--- a/kernel/sched/sched.h
+++ b/kernel/sched/sched.h
@@ -71,6 +71,7 @@
 #include <linux/wait_bit.h>
 #include <linux/workqueue_api.h>
 #include <linux/delayacct.h>
+#include <linux/uek_kabi.h>
 
 #include <trace/events/power.h>
 #include <trace/events/sched.h>
@@ -470,6 +471,10 @@ struct cfs_bandwidth {
 	UEK_KABI_RESERVE(1)
 	UEK_KABI_RESERVE(2)
 #endif
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+	UEK_KABI_RESERVE(5)
+	UEK_KABI_RESERVE(6)
 };
 
 /* Task group related information */
@@ -881,6 +886,11 @@ struct rt_rq {
 	struct rq		*rq;
 	struct task_group	*tg;
 #endif
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 static inline bool rt_rq_is_runnable(struct rt_rq *rt_rq)
@@ -948,6 +958,11 @@ struct dl_rq {
 	 * by the GRUB algorithm.
 	 */
 	u64			bw_ratio;
+
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
 };
 
 #ifdef CONFIG_FAIR_GROUP_SCHED
diff --git a/net/devlink/devl_internal.h b/net/devlink/devl_internal.h
index 4eee580614b6..31578928f0b8 100644
--- a/net/devlink/devl_internal.h
+++ b/net/devlink/devl_internal.h
@@ -3,6 +3,7 @@
  * Copyright (c) 2016 Jiri Pirko <jiri@mellanox.com>
  */
 
+#include <linux/uek_kabi.h>
 #include <linux/device.h>
 #include <linux/etherdevice.h>
 #include <linux/mutex.h>
@@ -60,6 +61,12 @@ struct devlink {
 	struct rcu_work rwork;
 	struct devlink_rel *rel;
 	struct xarray nested_rels;
+	
+	UEK_KABI_RESERVE(1)
+	UEK_KABI_RESERVE(2)
+	UEK_KABI_RESERVE(3)
+	UEK_KABI_RESERVE(4)
+
 	char priv[] __aligned(NETDEV_ALIGN);
 };
 
-- 
2.47.3

